plugin/trino-exchange/src/main/java/io/trino/plugin/exchange/s3/S3ExchangeStorage.java

```diff
@@ -0,0 +1,686 @@
+                throw e;
+            }
+
+            try {
+                uploadId.ifPresent(this::finishUpload);
+            }
+            catch (RuntimeException e) {
+                abortUploadSuppressed(e);
+                throw new IOException(e);
+            }
+        }
+
+        private void flushBuffer(boolean finished)
+                throws IOException
+        {
+            try {
+                waitForPreviousUploadFinish();
+            }
+            catch (IOException e) {
+                failed = true;
+                abortUploadSuppressed(e);
+                throw e;
+            }
+
+            // skip multipart upload if there would only be one part
+            if (finished && uploadId.isEmpty()) {
+                InputStream in = new ByteArrayInputStream(buffer, 0, bufferSize);
+
+                ObjectMetadata metadata = new ObjectMetadata();
+                metadata.setContentLength(bufferSize);
+                metadata.setContentMD5(getMd5AsBase64(buffer, 0, bufferSize));
+
+                PutObjectRequest request = new PutObjectRequest(bucketName, key, in, metadata);
+                sseCustomerKey.ifPresent(request::withSSECustomerKey);
+
+                try {
+                    s3.putObject(request);
+                    return;
+                }
+                catch (AmazonServiceException e) {
+                    throw new IOException(e);
+                }
+            }
+
+            // The multipart upload API only accept the last part to be less than 5MB
+            if (bufferSize == buffer.length || (finished && bufferSize > 0)) {
+                byte[] data = buffer;
+                int length = bufferSize;
+                this.buffer = new byte[buffer.length];
+                bufferSize = 0;
+
+                inProgressUploadFuture = uploadExecutor.submit(() -> uploadPage(data, length));
+            }
+        }
+
+        private UploadPartResult uploadPage(byte[] data, int length)
+        {
+            if (uploadId.isEmpty()) {
+                uploadId = Optional.of(uploadIdFactory.get());
+            }
+
+            currentPartNumber++;
+            UploadPartRequest uploadRequest = new UploadPartRequest()
+                    .withBucketName(bucketName)
+                    .withKey(key)
+                    .withUploadId(uploadId.get())
+                    .withPartNumber(currentPartNumber)
+                    .withInputStream(new ByteArrayInputStream(data, 0, length))
+                    .withPartSize(length)
+                    .withMD5Digest(getMd5AsBase64(data, 0, length));
+            sseCustomerKey.ifPresent(uploadRequest::withSSECustomerKey);
+
+            UploadPartResult partResult = s3.uploadPart(uploadRequest);
+            parts.add(partResult);
+            return partResult;
+        }
+
+        private static String getMd5AsBase64(byte[] data, int offset, int length)
+        {
+            @SuppressWarnings("deprecation")
```

Comment:
can we use non-deprecated api?

Reply:
I think it's because AWS S3 only provides MD5 hash of the content as a integrity validation method.